lstchain/mc/sensitivity.py

```diff
@@ -106,86 +115,138 @@ def bin_definition(gb, tb):
-                  mc_par_p['cone'], mc_par_p['area_sim'], proton_par['f0'], proton_par['e0'])
+    rate_p = rate(mc_par_p['emin'], mc_par_p['emax'], mc_par_p['sp_idx'],
+                     mc_par_p['cone'], mc_par_p['area_sim'],
+                     proton_par['f0'], proton_par['e0'])
 
 
     w_g = weight(mc_par_g['emin'], mc_par_g['emax'], mc_par_g['sp_idx'],
-                 crab_par['alpha'], rate_g, mc_par_g['sim_ev'], crab_par['e0'])
+                    crab_par['alpha'], rate_g,
+                    mc_par_g['sim_ev'], crab_par['e0'])
 
     w_p = weight(mc_par_p['emin'], mc_par_p['emax'], mc_par_p['sp_idx'],
-                 proton_par['alpha'], rate_p, mc_par_p['sim_ev'], proton_par['e0'])
+                    proton_par['alpha'], rate_p,
+                    mc_par_p['sim_ev'], proton_par['e0'])
 
 
     e_reco_gw = ((e_reco_g / crab_par['e0'])**(crab_par['alpha'] - mc_par_g['sp_idx'])) \
-        * w_g
+                * w_g
     e_reco_pw = ((e_reco_p / proton_par['e0'])**(proton_par['alpha'] - mc_par_g['sp_idx'])) \
-        * w_p
+                * w_p
 
     # Arrays to contain the number of gammas and hadrons for different cuts
+
     final_gamma = np.ndarray(shape=(eb, gb, tb))
     final_hadrons = np.ndarray(shape=(eb, gb, tb))
 
-    g, t = bin_definition(gb, tb)
+
+    #Weight events and count nº of events per bin:
 
     for i in range(0,eb):  # binning in energy
         for j in range(0,gb):  # cut in gammaness
             for k in range(0,tb):  # cut in theta2
-                eg_w_sum = np.sum(e_reco_gw[(e_reco_g < E[i+1].to_value()) & (e_reco_g > E[i].to_value()) \
-                                         & (gammaness_g > g[j]) & (theta2_g < t[k])])
+                eg_w_sum = np.sum(e_reco_gw[(e_reco_g < E[i+1]) & (e_reco_g > E[i]) \
+                                            & (gammaness_g > g[j]) & (theta2_g < t[k])])
+
+                ep_w_sum = np.sum(e_reco_pw[(e_reco_p < E[i+1]) & (e_reco_p > E[i]) \
+                                            & (gammaness_p > g[j]) & (theta2_p < t[k])])
 
-                ep_w_sum = np.sum(e_reco_pw[(e_reco_p < E[i+1].to_value()) & (e_reco_p > E[i].to_value()) \
-                                         & (gammaness_p > g[j]) & (theta2_p < t[k])])
-            
                 final_gamma[i][j][k] = eg_w_sum * obstime
                 final_hadrons[i][j][k] = ep_w_sum * obstime
-                    
-    sens = calculate_sensitivity(final_gamma.to_value(), final_hadrons.to_value(), 1/noff)
-    
+
+    sens = calculate_sensitivity(final_gamma, final_hadrons, 1/noff)
+
+    #Avoid bins which are empty or have too few events:
+
+    min_num_events = 10 #Minimum number of gamma and proton events in a bin to be taken into
+    #account for minimization
+
+    for i in range(0, eb):
+        for j in range(0, gb):
+            for k in range(0, tb):
+                if np.isnan(sens[i][j][k]) or np.isinf(sens[i][j][k]) \
```

Comment:
Can be simplified imo with 
```
conditions = (not np.isfinite(sens[i,j,k])) or (sens[i,j,k]<=0) or (final_gamma[i,j,k] < min_num_events) or (final_hadrons[i,j,k] < min_num_events) or (not final_gamma[i,j,k] > final_hadrons[i,j,k] * 0.05)
if cond:
    sens[i][j][k] = 1e100
```
why 1e100 and not `np.nan` stating clearly this is not a good value ?

Reply:
I set it to a huge value because later, the loop to chose the bin with minimum sensitivity was getting crazy with nans and infs.

Reply:
In this case, the isfinite condition should be verified in the loop imo.

Reply:
Finally got to change np.nan for np.inf because np.argmin() was taking "nan" as a minimum, and np.nanarming() doesn't like when an array is full of nans...

Reply:
I would personally prefer to have `nan` values when the sensitivity has not been calculated properly but if it's working like that, I think we can go with it.